pkg/core/block/block.go

```diff
@@ -112,29 +123,60 @@ func (b *Block) Trim() ([]byte, error) {
 	buf.WriteB(1)
 	b.Script.EncodeBinary(buf.BinWriter)
 
-	buf.WriteVarUint(uint64(len(b.Transactions)))
+	buf.WriteVarUint(uint64(len(b.Transactions)) + 1)
+	hash := b.ConsensusData.Hash()
+	hash.EncodeBinary(buf.BinWriter)
+
 	for _, tx := range b.Transactions {
 		h := tx.Hash()
 		h.EncodeBinary(buf.BinWriter)
 	}
+
+	b.ConsensusData.EncodeBinary(buf.BinWriter)
 	if buf.Err != nil {
 		return nil, buf.Err
 	}
+
 	return buf.Bytes(), nil
 }
 
 // DecodeBinary decodes the block from the given BinReader, implementing
 // Serializable interface.
 func (b *Block) DecodeBinary(br *io.BinReader) {
 	b.Base.DecodeBinary(br)
-	br.ReadArray(&b.Transactions)
+	contentsCount := br.ReadVarUint()
+	if contentsCount == 0 {
+		br.Err = errors.New("invalid block format")
+		return
+	}
+	b.ConsensusData.DecodeBinary(br)
+	txes := make([]*transaction.Transaction, contentsCount-1)
+	for i := 0; i < int(contentsCount)-1; i++ {
+		tx := new(transaction.Transaction)
+		tx.DecodeBinary(br)
+		txes[i] = tx
+	}
+	b.Transactions = txes
+	merkle, err := b.computeMerkleTree()
```

Comment:
I don't really like this check here, we have it in `Verify()`, but at the same time treating Merkle tree hash mismatch as a decoding error can be seen as a part of the protocol.

BTW, what we're currently completely missing is transaction duplication check.

Reply:
Added check to `block.DecodeBinary`. Should we add this check somewhere else?

Reply:
Let's add it to `Verify()` also (maybe make a bit more common code). It's safer for us to have it everywhere for now and we can change it later if need be.

Reply:
Added check to `Verify()`. Now we call `Verify()` after block was decoded.

Reply:
OK, looks good given that we have no other checks in `Verify` now.